spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala

```diff
@@ -0,0 +1,366 @@
+    val embeddingInput = SelectTable(1).inputs(split)
+    val embeddingOutput = SelectTable(2).inputs(split)
+
+    val encoderOutput = encode(embeddingInput, attentionBias)
+    val outputNode = decode(embeddingOutput, encoderOutput, attentionBias)
+    Graph(Array(inputNode, targetNode), outputNode)
+  }
+
+  private def buildLM(): Module[T] = {
+    val inputNode = Input()
+    val constantValue = math.sqrt(hiddenSize)
+    val embeddingInput = MulConstant(constantValue).inputs(
+      LookupTable[T](vocabSize, hiddenSize).inputs(inputNode))
+    val outputNode = decode(embeddingInput)
+    Graph(inputNode, outputNode)
+  }
+
+  private[nn] def encode(inputs: ModuleNode[T], attentionBias: ModuleNode[T]): ModuleNode[T] = {
+    // Prepare inputs to the layer stack by adding positional encodings and
+    // applying dropout.
+    val position = new PositionEncode().inputs(inputs)
+    val encoderInput = CAddTable().inputs(inputs, position)
+    val decoderInputDrop = if (train) {
+      val postDropOut = Dropout(1- postprocessDropout)
+      postDropOut.inputs(encoderInput)
+    } else encoderInput
+
+    encodeStack(numHiddenlayers, decoderInputDrop, attentionBias)
+  }
+
+  private[nn] def decode(targets: ModuleNode[T],
+                     encoderOutput: ModuleNode[T] = null,
+                     attentionBias: ModuleNode[T] = null): ModuleNode[T] = {
+    val decoderInput = new PositionEncodeWithShift().inputs(targets)
+    val decoderSelfAttentionBias = new SelfAttentionMask().inputs(targets)
+
+    val decoderInputDrop = if (train) {
+      val postDropOut = Dropout(1- postprocessDropout)
+      postDropOut.inputs(decoderInput)
+    } else decoderInput
+
+    decodeStack(numHiddenlayers, decoderInputDrop,
+      decoderSelfAttentionBias, encoderOutput, attentionBias)
+  }
+
+
+  private[nn] def encodeStack(numLayers: Int,
+                              encoderInput: ModuleNode[T],
+                              attentionBias: ModuleNode[T]): ModuleNode[T] = {
+    decodeStack(numLayers, encoderInput, attentionBias, preName = "encode")
+  }
+
+  private[nn] def decodeStack(numLayers: Int,
+                              decoderInput: ModuleNode[T],
+                              decoderSelfAttentionBias: ModuleNode[T],
+                              encoderOutputs: ModuleNode[T] = null,
+                              attentionBias: ModuleNode[T] = null,
+                              preName: String = "decode"): ModuleNode[T] = {
+    var input = decoderInput
+    var i = 0
+    while (i < numLayers) {
+      val selfAttention = new Attention[T](hiddenSize, numHeads, attentionDropout)
+      val selfAttentionModel = prePostProcessingSelfAttention(
+        selfAttention, input, decoderSelfAttentionBias,
+        s"${preName}_self_attention_${i}")
+      input = selfAttentionModel
+
+      if (encoderOutputs != null && attentionBias != null) {
+        val encdecAttention = new Attention[T](hiddenSize, numHeads, attentionDropout)
```

Comment:
encdec? typo ？

Reply:
no, it means attention with encode output and decode input